src/kudu/consensus/leader_election.cc

```diff
@@ -138,6 +138,113 @@ bool VoteCounter::AreAllVotesIn() const {
   return GetTotalVotesCounted() == num_voters_;
 }
 
+void FlexibleVoteCounter::FetchTopologyInfo(
+    std::map<std::string, int>& voter_distribution,
+    std::map<std::string, int>& data_quorum_requirement) {
+  voter_distribution.clear();
+  data_quorum_requirement.clear();
+  // TODO(ritwikyadav): Populate the out params by reading config.
+}
+
+FlexibleVoteCounter::FlexibleVoteCounter(RaftConfigPB config)
+  : VoteCounter(0, 0),
+    config_(std::move(config)) {
+  num_voters_ = 0;
+  yes_vote_count_.clear();
+  no_vote_count_.clear();
+  FetchTopologyInfo(voter_distribution_, data_quorum_requirement_);
+  for (auto regional_vote_count : voter_distribution_) {
+    CHECK_GT(regional_vote_count.second, 0);
+    num_voters_ += regional_vote_count.second;
+    yes_vote_count_.emplace(regional_vote_count.first, 0);
+    no_vote_count_.emplace(regional_vote_count.first, 0);
+  }
+
+  CHECK_GT(num_voters_, 0);
+
+  bool non_empty_quorum = false;
+  for (auto regional_quorum_count : data_quorum_requirement_) {
+    auto total_voters = voter_distribution_.find(regional_quorum_count.first);
+    CHECK(total_voters != voter_distribution_.end());
+    CHECK_LE(regional_quorum_count.second, total_voters->second);
+    if (regional_quorum_count.second > 0) {
+      non_empty_quorum = true;
+    }
+  }
+  CHECK(non_empty_quorum);
+
+  for (const RaftPeerPB& peer : config_.peers()) {
+    if (peer.member_type() == RaftPeerPB::VOTER) {
+      uuid_to_region_.emplace(peer.permanent_uuid(), peer.attrs().region());
+    }
+  }
+}
+
+Status FlexibleVoteCounter::RegisterVote(
+    const std::string& voter_uuid, ElectionVote vote, bool* is_duplicate) {
+  Status s = VoteCounter::RegisterVote(voter_uuid, vote, is_duplicate);
+  if (ContainsKey(uuid_to_region_, voter_uuid)) {
+    const std::string region = uuid_to_region_[voter_uuid];
+    switch (vote) {
+      case VOTE_GRANTED:
+        InsertIfNotPresent(&yes_vote_count_, region, 0);
+        yes_vote_count_[region]++;
+        break;
+      case VOTE_DENIED:
+        InsertIfNotPresent(&no_vote_count_, region, 0);
+        no_vote_count_[region]++;
+        break;
+    }
+  } else {
+    return Status::InvalidArgument(
+        Substitute("UUID {$0} not present in config.", voter_uuid));
+  }
+  return s;
+}
+
+std::pair<bool, bool> FlexibleVoteCounter::GetQuorumState() const {
+  bool quorum_satisfied = true;
```

Comment:
every ResponseFromPeer will call this function. Maybe we can think of making this incremental